src/LambdaCube/Compiler/Statements.hs

```diff
@@ -112,12 +112,12 @@ compileStmt lhs compilegt ds = \case
         alts -> compileStmt lhs compileGuardTrees' [TypeAnn n t] alts
     fs@(FunAlt n vs _: _) -> case groupBy ((==) `on` fst) [(length vs, n) | FunAlt n vs _ <- fs] of
         [gs@((num, _): _)]
-          | num == 0 && length gs > 1 -> fail $ "redefined " ++ sName n ++ ":\n" ++ show (vcat $ pShow . sourceInfo . snd <$> gs)
+          | num == 0 && length gs > 1 -> error $ "redefined " ++ sName n ++ ":\n" ++ show (vcat $ pShow . sourceInfo . snd <$> gs)
```

Comment:
Does not this change the semantics?
The `error` is more lazy than `fail` AFAIK.

Reply:
Disclaimer: I'm far from an expert in Haskell (I have a few hobby projects written in it)…

I _think_ if I understood the MonadFail Proposal correctly, `fail` for Monads that didn't have a "reasonable" `fail` implementation and will not become MonadFail in the new proposal, **was already implemented by calling `error`** (which is part of why they are removing `fail` for those Monads).

If someone who actually is familiar with the internals of the MFP could confirm this, that would be good to know for sure.